pages/blog/posts/understanding-lexical-dynamic-scopes.md

```diff
@@ -0,0 +1,252 @@
+---
+
+Most of the [keywords](https://json-schema.org/learn/glossary#keyword) defined
+by the JSON Schema organization can be either evaluated on their own, or by
+only considering the values of their adjacent keywords. For example, the `type`
+keyword is independent of any other keywords, while the `minContains` and
+`maxContains` keywords depend on the `contains` keyword defined in the same
+schema object.
+
+> If you want to learn more about keyword dependencies, check out the [Static
+> Analysis of JSON Schema](schema-static-analysis) article by [Greg
+> Dennis](https://www.linkedin.com/in/gregdennis/).
+
+However, there is a small set of keywords whose evaluation depend on the
+*scope* they are in. These keywords are `$ref`, `$dynamicRef`,
+`unevaluatedItems`, and `unevaluatedProperties`. Additionally, there is also a
+set of keywords that *affect* the scope they are declared in. These keywords
+are `$id`, `$schema`, `$anchor`, and `$dynamicAnchor`.
+
+JSON Schema defines two types of scopes: the *lexical scope* and the *dynamic
+scope*. Understanding how these scopes work is essential for mastering some of
+the most advanced (and often confusing!) features of JSON Schema, such as
+dynamic referencing.
+
+Lexical Scope
+-------------
+
+JSON Schema is a circular data structure. What we mean by this is that as a
+schema author, you may use *applicator* keywords that take other
+[schemas](https://json-schema.org/learn/glossary#schema) as arguments,
+effectively creating a tree of schemas. 
+
+Consider the following example. On the left, a JSON Schema consisting of 7
+[subschemas](https://json-schema.org/learn/glossary#subschema) (declared using
+the `properties` and `prefixItems` applicators). On the right, a tree
+representation of the schema, where each subschema corresponds to a node in the
+tree:
+
+![Thinking of a JSON Schema as a tree structure](/img/posts/2024/understanding-lexical-dynamic-scopes/tree.webp)
+
+Thinking of a schema as a tree structure greatly helps in understanding lexical
+scopes. Under this analogy, the lexical scope of a schema consists of the
+*ancestors* of such node, including the node itself. In other words, the
+lexical scope of a schema consists of its parent schemas including the schema
+itself, but not its subschemas.
+
+Referring back to our previous example, the lexical scope of the subschema at
+`/properties/bar/prefixItems/0` (highlighted in red) consists of the set of its
+parent schemas. The subschemas that are not part of its lexical scope have been
+grayed out:
+
+![The lexical scope of a JSON Schema represented as a tree structure](/img/posts/2024/understanding-lexical-dynamic-scopes/ancestor.webp)
+
+In comparison to the dynamic scope, the lexical scope of any subschema can be
+statically determined without taking
+[instances](https://json-schema.org/learn/glossary#instance) into account, just
+as we did here.
```

Comment:
Hm, I guess I'm confused then. The spec seems to only mention the evaluation path + instance with respects to the dynamic scope, and never with respects to the lexical scope:

> Other keywords may take into account the dynamic scope that exists during the evaluation of a schema, typically together with an instance document.

which made me think the distinction only applied to the dynamic scope

Reply:
There is also this part of the spec:

> The lexical scope of a keyword is determined by the nested JSON data structure of objects and arrays.

In conjunction with:

> Other keywords may take into account the dynamic scope that exists during the evaluation of a schema, typically together with an instance document. The outermost dynamic scope is the schema object at which processing begins, even if it is not a schema resource root. The path from this root schema to any particular keyword (that includes any "$ref" and "$dynamicRef" keywords that may have been resolved) is considered the keyword's "validation path."

Which made me think the lexical scope is about the structure of the schema while the dynamic scope was about the validation/evaluation path

Reply:
Also on this part:

> A keyword on the far side of that reference that resolves information through the dynamic scope will consider the originating side of the reference to be their dynamic parent, rather than examining the local lexically enclosing parent.

The spec seems to say that the origin of a reference (and note it never says whether across schema resources or not) seems to only be taken into account for the dynamic scope, which is the reason I didn't consider the lexical scope as the evaluation path too, as you are pointing out on this example.

Anyway, not trying to say you are wrong, but trying to echo out loud what parts of the spec are confusing me

Reply:
I don't think it makes sense to say that the scope is the entire schema resource. Take this javascript as an analogous scope example.

```javascript
function foo() {
  let a;
  function bar() {
    let b;
    function baz() {
      let c;
    }
  }
}
```

When in the `bar` function, `a` and `b` are in scope. Anything within and lexically above the current block is in scope. If the entire document was in scope, it would mean you would mean `c` is also in scope from `bar`, which I don't think makes sense.

Reply:
Tying back to @jdesrosiers 's comment here (https://github.com/json-schema-org/website/pull/274#discussion_r1465740269), if we say a schema resource is analogous to a class and that a subschema is analogous to a method, then maybe what @Relequestual is saying makes sense:

- The lexical scope is the class
- When you are in any of its methods (i.e. subschemas), your scope is indeed the entire class, as you can access its members and other methods

Reply:
> I don't think it makes sense to say that the scope is the entire schema resource.

Understood, however an embedded schema resource's contense is not included in the containing schema resource. `baz` cann't see `a` or `b`.

A better analogy is individual functions in separate classes, all imported into the same class, then re-exported. They have their own lexical scope, but can be embedded in other container classes for ease of use.

Reply:
> When you are in any of its methods (i.e. subschemas), your scope is indeed the entire class, as you can access its members and other methods

Not the entire class. When in a method, you can access your own local variable, but not the local variables of another method. Just like when evaluating `additionalProperties` you only have access to the `properties` keyword in the same subchema, not the `properties` keyword in another subschema. You can access the resource-scoped variables from the entire class, but you can only access method-scope variables from the method you're in.

> A better analogy is individual functions in separate classes, all imported into the same class, then re-exported. They have their own lexical scope, but can be embedded in other container classes for ease of use.

I would compare embedded schemas to inner classes (a class defined inside a class). I'm not sure I entirely follow your analogy, but I think it's effectively the same idea?